### ISSUE TITLE
RobotWars koden

### ISSUE THREAD

**Issue by AlbinH1** [AlbinH1/RobotWars#1]
#include <Servo.h> // Include servo library

Servo servoLeft; // Declare left and right servos

Servo servoRight;

byte wLeftOld; // Previous loop whisker values

byte wRightOld;

byte counter; // For counting alternate corners

void setup() // Built-in initialization block

{

pinMode(10, INPUT); // Set right whisker pin to input

pinMode(11, INPUT); // Set left whisker pin to input

servoLeft.attach(12); // Attach left signal to pin 13

servoRight.attach(13); // Attach right signal to pin 12

wLeftOld = 0; // Init. previous whisker states

wRightOld = 1;

counter = 0; // Initialize counter to 0

pinMode(9, INPUT); pinMode(8, OUTPUT); // Left IR LED & Receiver

pinMode(3, INPUT); pinMode(2, OUTPUT); // Right IR LED & Receiver

Serial.begin(9600);

}

void loop() // Main loop auto-repeats
{

//int irLeft = 1; // Declare IR variables and

//int irRight = 1; // initialize both to 1

//while((irLeft == 1) && (irRight == 1)) {// Wait for detection

int irLeft = irDetect(8, 9, 45000); // Check for object on left

int irRight = irDetect(2, 3, 44000); // Check for object on right
//}

Serial.println(irLeft); 
Serial.println(irRight);
delay(100);

if((irLeft == 0) && (irRight == 0)) // If both sides detect

{

backward(1500); // Back up 1,5 second

turnLeft(100); // Turn left about 15 degrees

}

else if(irLeft == 0) // If only left side detects

{

turnRight(50); // Turn right about 7,5 degrees

}

else if(irRight == 0) // If only right side detects

{

turnLeft(50); // Turn left about 7,5 degrees

}

else // Otherwise, no IR detected

{

forward(40); // Forward 1/25 of a second

}

}
int irDetect(int irLedPin, int irReceiverPin, long frequency) {

tone(irLedPin, frequency, 8); // IRLED 38 kHz for at least 1 ms

delay(1); // Wait 1 ms

int ir = digitalRead(irReceiverPin); // IR receiver -> ir variable

delay(1); // Down time before recheck

return ir; // Return 1 no detect, 0 detect

}

void maneuver(int speedLeft, int speedRight, int msTime)

{

// speedLeft, speedRight ranges: Backward Linear Stop Linear Forward

// -200 -100......0......100 200

servoLeft.writeMicroseconds(1500 + speedLeft); // Set left servo speed

servoRight.writeMicroseconds(1500 - speedRight); // Set right servo speed

if(msTime==-1) // if msTime = -1

{

servoLeft.detach(); // Stop servo signals
servoRight.detach();
}

delay(msTime); // Delay for msTime

// Corner Escape

byte wLeft = digitalRead(11); // Copy right result to wLeft

byte wRight = digitalRead(10); // Copy left result to wRight

if(wLeft != wRight) // One whisker pressed?

{ // Alternate from last time?

if ((wLeft != wLeftOld) && (wRight != wRightOld))

{

counter++; // Increase count by one

wLeftOld = wLeft; // Record current for next rep

wRightOld = wRight;

if(counter == 4) // Stuck in a corner?

{

wLeft = 0; // Set up for U-turn

wRight = 0;

counter = 0; // Clear alternate corner count

}

}

else // Not alternate from last time

{

counter = 0; // Clear alternate corner count

}

}

// Whisker Navigation

if((wLeft == 0) && (wRight == 0)) // If both whiskers contact

{

backward(1000); // Back up 1 second

turnLeft(800); // Turn left about 120 degrees

}

else if(wLeft == 0) // If only left whisker contact

{

backward(1000); // Back up 1 second

turnRight(400); // Turn right about 60 degrees

}

else if(wRight == 0) // If only right whisker contact

{

backward(1000); // Back up 1 second

turnLeft(400); // Turn left about 60 degrees

}

else // Otherwise, no whisker contact

{

forward(20); // Forward 1/50 of a second

}

}

void forward(int t) // Forward function

{

servoLeft.writeMicroseconds(1700); // Left wheel counterclockwise

servoRight.writeMicroseconds(1300); // Right wheel clockwise

delay(t); // Maneuver for time ms

}

void turnLeft(int t) // Left turn function

{

servoLeft.writeMicroseconds(1300); // Left wheel clockwise

servoRight.writeMicroseconds(1300); // Right wheel clockwise

delay(t); // Maneuver for time ms

}

void turnRight(int t) // Right turn function

{

servoLeft.writeMicroseconds(1700); // Left wheel counterclockwise

servoRight.writeMicroseconds(1700); // Right wheel counterclockwise

delay(t); // Maneuver for time ms

}

void backward(int t) // Backward function

{

servoLeft.writeMicroseconds(1300); // Left wheel clockwise

servoRight.writeMicroseconds(1700); // Right wheel counterclockwise

delay(t); // Maneuver for time ms

}
